File: web/modules/guided-exploration/types/conversation.ts
/**
 * Conversation Types
 * Represents messages and conversations within leaf nodes
 */

import type { Citation, SubtopicContent } from './content';

export type MessageRole = 'system' | 'user' | 'assistant';
export type MessageType =
  | 'initial_content'
  | 'followup'
  | 'analysis'
  | 'system_message';

// ============ Session Messages ============

export type SessionMessageType =
  | 'user'
  | 'assistant'
  | 'exploration_start'
  | 'topic_directions';

export interface PartyStanceSummary {
  partyId: string;
  summary: string;
}

export interface ExplorationOverview {
  introParagraph: string;
  partySummaries: PartyStanceSummary[];
}

export interface SessionMessage {
  /** Unique message identifier */
  id: string;
  /** Type of session message */
  type: SessionMessageType;
  /** Message content (null for structured messages like exploration_start) */
  content: string | null;
  /** Citations for this message (for quick summaries) */
  citations?: Citation[];
  /** ID of the exploration if this is an exploration_start message */
  explorationId?: string;
  /** The query that triggered the exploration */
  explorationQuery?: string;
  /** Topic direction choices (for topic_directions type) */
  directions?: Array<{
    id: string;
    name: string;
    hook: string;
    suggestedQuestion: string;
  }>;
  /** Query ID for direction choice tracking */
  directionsQueryId?: string;
  /** Names of directions the user selected (set after submission) */
  selectedDirections?: string[];
  /** When the message was created */
  timestamp: string;
}

export interface Message {
  /** Unique message identifier */
  id: string;
  /** Who sent the message */
  role: MessageRole;
  /** Type of message */
  type: MessageType;
  /** Message content - string for followups, structured for initial content */
  content: string | SubtopicContent;
  /** Citations for followup messages */
  citations?: Citation[];
  /** When the message was created */
  timestamp: string;
}

export interface Conversation {
  /** Leaf node identifier this conversation belongs to */
  leafId: string;
  /** All messages in this conversation */
  messages: Message[];
}

File: ai-backend/src/guided_exploration/models/conversation.py
# SPDX-FileCopyrightText: 2025 wahl.chat
#
# SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0

"""Models for conversations within leaf nodes in guided exploration."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field

from src.guided_exploration.models.content import Analysis, Citation, SubtopicContent


class MessageType(str, Enum):
    """Type of message in a conversation."""

    INITIAL_CONTENT = "initial_content"
    FOLLOWUP = "followup"
    ANALYSIS = "analysis"


class MessageRole(str, Enum):
    """Role of the message author."""

    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"


class TopicSwitchProposal(BaseModel):
    """A neighbour-leaf switch proposal emitted alongside an assistant turn.

    Lives at the model layer (not the agent layer) because it's persisted
    on the Message and read back by both the leaf-followup generator and
    the leaf conversation handler on subsequent turns.
    """

    target_node_id: str = Field(
        ..., description="ID of the sibling leaf the question fits better."
    )
    target_node_name: str = Field(
        ..., description="Display name of the target leaf."
    )
    reason: str = Field(
        ...,
        description=(
            "Short German message shown next to the switch action — "
            "why the other leaf fits better."
        ),
    )


class Message(BaseModel):
    """A single message in a conversation."""

    id: str = Field(..., description="Unique message identifier")
    role: MessageRole = Field(..., description="Who sent the message")
    type: MessageType = Field(..., description="Type of message content")
    content: str | SubtopicContent | Analysis = Field(
        ...,
        description="String for followups, SubtopicContent for initial, Analysis for analysis",
    )
    citations: list[Citation] = Field(
        default_factory=list,
        description="Citations for this message (persisted for reload)",
    )
    suggested_followups: list[str] = Field(
        default_factory=list,
        description=(
            "Follow-up chips emitted alongside this assistant turn. "
            "Persisted so later turns can see what was offered."
        ),
    )
    closure_ready: bool = Field(
        default=False,
        description=(
            "Whether the leaf was deemed substantially explored at this "
            "turn. Persisted so a later turn can see prior closure "
            "offerings — if the user kept exploring after a closure pill, "
            "the next turn should not re-offer closure unless something "
            "substantially new has happened."
        ),
    )
    topic_switch_proposal: TopicSwitchProposal | None = Field(
        default=None,
        description=(
            "Neighbour-leaf switch proposal emitted alongside this "
            "assistant turn, if any. Persisted so the same switch is not "
            "offered again on every subsequent turn."
        ),
    )
    timestamp: datetime = Field(..., description="When the message was sent")


class Conversation(BaseModel):
    """Conversation for a leaf node."""

    leaf_id: str = Field(
        ..., description="ID of the leaf node, e.g., 'housing.rent-control'"
    )
    messages: list[Message] = Field(
        default_factory=list, description="Messages in the conversation"
    )
